app/com/linkedin/drelephant/tuning/IPSOManager.java

```diff
@@ -0,0 +1,357 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.linkedin.drelephant.tuning;
+
+import com.linkedin.drelephant.util.MemoryFormatUtils;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import models.JobDefinition;
+import models.JobSuggestedParamSet;
+import models.JobSuggestedParamValue;
+import models.TuningAlgorithm;
+import models.TuningParameter;
+import org.apache.log4j.Logger;
+import models.AppHeuristicResult;
+import models.AppHeuristicResultDetails;
+import models.AppResult;
+import models.TuningParameterConstraint;
+import org.apache.commons.io.FileUtils;
+
+import static com.linkedin.drelephant.mapreduce.heuristics.CommonConstantsHeuristic.*;
+import static java.lang.Double.*;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+
+/*
+This Implements IPSO in AutoTuning Framework
+
+ */
+public class IPSOManager implements AutoTuningOptimizeManager {
+  private static final Logger logger = Logger.getLogger(IPSOManager.class);
+  private Map<String, Map<String, Double>> usageDataGlobal = null;
+  private String functionTypes[] = {"map", "reduce"};
+
+  enum USAGE_COUNTER_SCHEMA {USED_PHYSICAL_MEMORY, USED_VIRTUAL_MEMORY, USED_HEAP_MEMORY}
+
+  @Override
+  public void intializePrerequisite(TuningAlgorithm tuningAlgorithm, JobSuggestedParamSet jobSuggestedParamSet) {
+    logger.info(" Intialize Prerequisite ");
+    setDefaultParameterValues(tuningAlgorithm, jobSuggestedParamSet);
+  }
+
+  private void setDefaultParameterValues(TuningAlgorithm tuningAlgorithm, JobSuggestedParamSet jobSuggestedParamSet) {
+    List<TuningParameter> tuningParameters =
+        TuningParameter.find.where().eq(TuningParameter.TABLE.tuningAlgorithm, tuningAlgorithm).findList();
+    for (TuningParameter tuningParameter : tuningParameters) {
+      TuningParameterConstraint tuningParameterConstraint = new TuningParameterConstraint();
+      tuningParameterConstraint.jobDefinition = jobSuggestedParamSet.jobDefinition;
+      tuningParameterConstraint.tuningParameter = tuningParameter;
+      tuningParameterConstraint.lowerBound = tuningParameter.minValue;
+      tuningParameterConstraint.upperBound = tuningParameter.maxValue;
+      tuningParameterConstraint.constraintType = TuningParameterConstraint.ConstraintType.BOUNDARY;
+      tuningParameterConstraint.paramName = tuningParameter.paramName;
+      tuningParameterConstraint.save();
+    }
+  }
+
+  @Override
+  public void extractParameterInformation(List<AppResult> appResults) {
```

Comment:
This code to extractParameterInformation may be simplified. May be create a map containing keys with concatenation of "Heuristic Name" + "Heuristic Detail Name" by iterating through all application and finding max for each key. Finally you can get them by just using "Heuristic Name" + "Heuristic Detail Name" as key.